rotkehlchen/externalapis/defillama.py

```diff
@@ -0,0 +1,297 @@
+        """
+        if options is None:
+            options = {}
+        url = f'https://coins.llama.fi/{module}/'
+        if subpath:
+            url += subpath
+
+        log.debug(f'Querying defillama: {url}?{urlencode(options)}')
+        try:
+            response = self.session.get(
+                f'{url}?{urlencode(options)}',
+                timeout=DEFAULT_TIMEOUT_TUPLE,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'Defillama API request failed due to {str(e)}') from e
+
+        if response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
+            self.last_rate_limit = ts_now()
+            msg = f'Got rate limited by Defillama querying {url}'
+            log.warning(msg)
+            raise RemoteError(message=msg, error_code=HTTPStatus.TOO_MANY_REQUESTS)
+
+        if response.status_code != 200:
+            msg = (
+                f'Defillama API request {response.url} failed with HTTP status '
+                f'code: {response.status_code}'
+            )
+            raise RemoteError(msg)
+
+        try:
+            decoded_json = json.loads(response.text)
+        except json.decoder.JSONDecodeError as e:
+            msg = f'Invalid JSON in Defillama response. {e}'
+            raise RemoteError(msg) from e
+
+        return decoded_json
+
+    def _get_asset_id(self, asset: AssetWithOracles) -> Optional[str]:
+        """
+        Create the id to be used in Defillama.
+        May raise:
+        - UnsupportedAsset
+        """
+        if asset.is_evm_token() is True:
+            asset = asset.resolve_to_evm_token()
+            return f'{str(asset.chain)}:{asset.evm_address}'
+
+        return f'coingecko:{asset.to_coingecko()}'
+
+    def query_current_price(
+            self,
+            from_asset: AssetWithOracles,
+            to_asset: AssetWithOracles,
+    ) -> Price:
+        """
+        Returns a simple price for from_asset to to_asset in Defillama.
+
+        May raise:
+        - RemoteError if there is a problem querying defillama
+        """
+        try:
+            coin_id = self._get_asset_id(from_asset)
+        except UnsupportedAsset:
+            log.warning(
+                f'Tried to query current price using Defillama from {from_asset} to '
+                f'{to_asset} but {from_asset} is not an EVM token and is not '
+                f'suppoorted by defillama',
+            )
+            return Price(ZERO)
+
+        result = self._query(
+            module='prices',
+            subpath=f'current/{coin_id}',
+        )
+
+        if 'coins' not in result or len(result['coins']) == 0:
+            log.warning(
+                f'Queried Defillama current price from {from_asset.identifier} '
+                f'to {to_asset.identifier}. But coins is not available in the result {result}',
+            )
+            return Price(ZERO)
+
+        coin_result_raw = result['coins'][coin_id]
+        try:
```

Comment:
I told you to try and abstract out and reuse duplicated code. The confidence code is probably the easiest target. But there is (probably?) more?